### ISSUE TITLE
Fail to run web server with Napa module

### ISSUE THREAD

**Issue by JiapengZhu** [Microsoft/napajs#139]
Hello
I am very new to use Napa module and just implemented a hello world web server using Napa module. This web server just prints "Hello Napa" in parallel by multiple threads. However, the problem is this web server cannot respond to client. I am not sure if the reason is Napa module does not support node express or http module or else. I know Napa has two zones, one is napa zone and the other is node zone, if napa does not support http or express module, how can I use node zone to compensate that? Thanks! 
I put my code below, 

```
var napa = require('napajs');
var app = require('express')();
var Promise = require('promise');

const NUMBER_OF_WORKERS = 4
	, PORT = 3005
	, HOST = '0.0.0.0';
	
var zone = napa.zone.create('zone', { workers: NUMBER_OF_WORKERS });
app.listen(PORT, HOST);

function startParellelRun(){	
	app.get('/', (req, res) => {
		run().then(() => {
			res.status(200).send('OK');
		}).catch(() => {
			res.status(500).send('ERROR');
		});		
	});
}

function run(){	
	var promises = [];	
	for(var i = 0; i < NUMBER_OF_WORKERS; i++){
		promises[i] = zone.execute(hello, [i]);		
	}	
	return Promise.all(promises)
			.then((values) => {			
				values.forEach((ele) => {
					console.log(ele.value);
				});				
			});								
}

function hello(i){
	return 'Hello Napa! I am Worker ' + i;
}

startParellelRun();

```

**Comment by tswaters** [comment]
I've encountered this myself.  

Originally tried using zone in an express route, letting a zone do work and return the result to the user - but was able to reduce it to the following case... it seems having any http server listening means promises from `execute` go unresolved.

```js
const zone1 = require('napajs').zone.create('zone1')
zone1.broadcast('function hello(arg) { return arg }')
zone1.execute('', 'hello', ['foo'], { timeout: 300 })
  .then(result => console.log(result.value))
  .catch(err => console.error(err))
// require('http').createServer().listen(3000, () => console.log('listening'))
```

* Run this, "foo" will be logged.
* Un-comment the last line and run again, you'll see "listening" and no "foo".

**Comment by tswaters** [comment]
Also, I was able to fix it in my node_modules copy by wrapping the `broadcast` and `execute` calls with `setImmediate`.  

I'd send a PR but I can't seem to get this to build locally... something about missing Visual C++ compiler (after having installed it *sigh*)

```diff
diff --git a/lib/zone/zone-impl.ts b/lib/zone/zone-impl.ts
index 3a6bad9..2aa8912 100644
--- a/lib/zone/zone-impl.ts
+++ b/lib/zone/zone-impl.ts
@@ -63,11 +63,13 @@ export class ZoneImpl implements zone.Zone {
 
         return new Promise<void>((resolve, reject) => {
             this._nativeZone.broadcast(source, (resultCode: number) => {
-                if (resultCode === 0) {
-                    resolve();
-                } else {
-                    reject("broadcast failed with result code: " + resultCode);
-                }
+                setImmediate(() => {
+                    if (resultCode === 0) {
+                        resolve();
+                    } else {
+                        reject("broadcast failed with result code: " + resultCode);
+                    }    
+                })
             });
         });
     }
@@ -77,13 +79,15 @@ export class ZoneImpl implements zone.Zone {
         
         return new Promise<zone.Result>((resolve, reject) => {
             this._nativeZone.execute(spec, (result: any) => {
-                if (result.code === 0) {
-                    resolve(new Result(
-                        result.returnValue,
-                        transport.createTransportContext(true, result.contextHandle)));
-                } else {
-                    reject(result.errorMessage);
-                }
+                setImmediate(() => {
+                    if (result.code === 0) {
+                        resolve(new Result(
+                            result.returnValue,
+                            transport.createTransportContext(true, result.contextHandle)));
+                    } else {
+                        reject(result.errorMessage);
+                    }
+                })
             });
         });
     }
```

**Comment by tswaters** [comment]
Ok, got my build going, but the tests fail with this change; turns out this won't work inside a thread due to `setImmediate` and various other timing-based mechanisms (process.nextTick, setTimeout, setInterval) not being implemented.  😢

**Comment by JiapengZhu** [comment]
@tswaters Thank you for your reply! 
```
const zone1 = require('napajs').zone.create('zone1')
zone1.broadcast('function hello(arg) { return arg }')
zone1.execute('', 'hello', ['foo'], { timeout: 300 })
  .then(result => console.log(result.value))
  .catch(err => console.error(err))
// require('http').createServer().listen(3000, () => console.log('listening'))
```
If you uncomment last line, you only see 'listening', because code is executed in async way. 'foo' printing is in callback function. Anyway, have you found a way to make http/express work in napajs? Thanks!

**Comment by JiapengZhu** [comment]
@tswaters if the way you create a new zone to handle responding to user with result using express/http module that works, can you show me a sample code for that? I am not sure how to let two zones communicate with each other.

**Comment by helloshuangzi** [comment]
Thanks for your feedback @JiapengZhu and @tswaters, and sorry for the late response due to ThanksGiving holiday.
@JiapengZhu. The quick answer to your question: Napa zone can't run most existing node modules, like http, by default. We have made node modules compatibility in our road map. 
Will look into your question about node zone execution.

**Comment by JiapengZhu** [comment]
@helloshuangzi Hello!
Have you considered dockerizing Napa.js module, creating a Napa.js docker image and pushing to public hub? Actually, i installed Napa.js in a docker container and publish a public port (3001:3005) that can be access by external. 
I used curl cmd to access Napa web server, 
`curl 192.168.1.178:3001/`
The client side kept waiting until 
`curl: (52) Empty reply from server `
or ctrl+C
meanwhile, server side printed out our expected message, 
```
Hello Napa! I am Worker 0
Hello Napa! I am Worker 1
Hello Napa! I am Worker 2
Hello Napa! I am Worker 3
```
I think we expect client side gets 200 status code and 'OK' message from server side.
Even though server side can print out something, client side gets nothing...
Does it happen the same thing when you test the code?

**Comment by helloshuangzi** [comment]
I haven't touched docker for napajs so far, but it's really a good idea to distribute napajs by docker. 
The web explorer can get 'OK' with the 2nd refresh in my client. You may try 'curl 192.168.1.178:3001/' in a two terminals to make the request twice. I tend to suspect it's the same problem w/ or w/o docker.

**Comment by tswaters** [comment]
> can you show me a sample code for that?

@JiapengZhu  I have a project here:  https://github.com/tswaters/react-thread-poc this includes a `nappajs-patch.diff`. I apply a patch manually to `napajs` to get it to work; basically, wrap the `resolve`s inside `./lib/zone/zone-impl.js` with a `setImmediate` calls and it works.

I was going to try and monkey patch zone to provide the ability to use a callback directly, so I can wrap it with `setImmeidate` myself, but because the `Result` class isn't exported from that file, it was going to be too much duplication for my liking.

@helloshuangzi I encountered the same issue on one of the other modules referenced in that project, `webworker-threads`. Same idea, wrapping a callback from a native module in `new Promise`,  the resolve or reject won't get called until something kicks the event loop... same workaround, wrap with `setImmediate`

The closest I found for an explanation was in this comment:  https://github.com/audreyt/node-webworker-threads/issues/123#issuecomment-254019552

And honestly I don't fully understand how native modules interact (or don't interact) with the nodejs event loop but it seems there is likely something missing in the cpp portion of this module to fix it "Properly"... in my eyes, `setImmediate` is just a workaround.

Also, for what it's worth, in terms of performance napajs blows everything else out of the water when I stress test that application.... if this was fixed I could definitely see myself using this module in production to speed up react apps that use server-side rendering (under load, using threads is ~300% faster)

**Comment by daiyip** [comment]
@tswaters, thanks for providing the patch. This is awesome. 
I guess it may also resolve #27 and #32. We should validate this patch and include it in next publish.

**Comment by tswaters** [comment]
Awesome!  Thanks so much for the quick action on this!

**Comment by daiyip** [comment]
@tswaters, you are welcome! Looking forward to seeing `napajs` used in your production. Please keep your feedback coming.

BTW, now 0.1.7 is ready to consume.

**Comment by daiyip** [comment]
Due to a publish issue, 0.1.7 had not picked up this change. 
We have published 0.1.8, please try. Sorry for inconvenience.
